libclamav/cvd.c

```
@@ -519,197 +522,185 @@ void cl_cvdfree(struct cl_cvd *cvd)
+            cli_errmsg("cl_cvdverify: Failed to set engine certs directory\n");
+            goto done;
+        }
+    }
 
-    ret = cli_cvdload(fs, engine, NULL, CL_DB_STDOPT | CL_DB_PUA, dbtype, file, 1);
+    if (!!cli_strbcasestr(file, ".cld")) {
+        dbtype = CVD_TYPE_CLD;
+    } else if (!!cli_strbcasestr(file, ".cud")) {
+        dbtype = CVD_TYPE_CUD;
+    }
+
+    ret = cli_cvdload(engine, NULL, CL_DB_STDOPT | CL_DB_PUA, dbtype, file, 1);
 
+done:
     cl_engine_free(engine);
-    fclose(fs);
+
     return ret;
 }
 
-cl_error_t cli_cvdload(FILE *fs, struct cl_engine *engine, unsigned int *signo, unsigned int options, unsigned int dbtype, const char *filename, unsigned int chkonly)
+cl_error_t cli_cvdload(struct cl_engine *engine, unsigned int *signo, unsigned int options, cvd_type dbtype, const char *filename, unsigned int chkonly)
 {
-    struct cl_cvd cvd, dupcvd;
-    FILE *dupfs;
+    cl_error_t status = CL_ECVD;
     cl_error_t ret;
     time_t s_time;
-    int cfd;
     struct cli_dbio dbio;
-    struct cli_dbinfo *dbinfo = NULL;
-    char *dupname;
+    struct cli_dbinfo *dbinfo  = NULL;
+    char *dupname              = NULL;
+    cvd_t *cvd                 = NULL;
+    cvd_t *dupcvd              = NULL;
+    FFIError *cvd_open_error   = NULL;
+    FFIError *cvd_verify_error = NULL;
+    char *signer_name          = NULL;
 
     dbio.hashctx = NULL;
 
     cli_dbgmsg("in cli_cvdload()\n");
 
-    /* verify */
-    if ((ret = cli_cvdverify(fs, &cvd, dbtype)))
-        return ret;
+    /* Open the cvd and read the header */
+    cvd = cvd_open(filename, &cvd_open_error);
+    if (!cvd) {
+        cli_errmsg("cli_cvdload: Can't open CVD file %s: %s\n", filename, ffierror_fmt(cvd_open_error));
+        goto done;
+    }
+
+    /* For actual .cvd files, verify the digital signature. */
+    if (dbtype == CVD_TYPE_CVD) {
+        if (!cvd_verify(
+                cvd,
+                engine->certs_directory,
+                false,
```

Consider if we want to make a config or commandline option to disable the older md5-based RSA checks, or simply let it fail if FIPS is enabled and the cert-based method doesn't work/isn't available.

I think it's okay either way.